packages/vite/src/node/utils.ts

```diff
@@ -534,15 +534,28 @@ export function isFileReadable(filename: string): boolean {
 
 /**
  * Delete every file and subdirectory. **The given directory must exist.**
- * Pass an optional `skip` array to preserve files in the root directory.
+ * Pass an optional `skip` array to preserve files under the root directory.
  */
-export function emptyDir(dir: string, skip?: string[]): void {
+export function emptyDir(dir: string, skip: string[] = []): void {
+  const nested: string[] = []
   for (const file of fs.readdirSync(dir)) {
-    if (skip?.includes(file)) {
+    const matched = skip.find((f) => f.startsWith(file))
+    if (matched) {
+      if (matched !== file) {
+        nested.push(`${dir}/${file}`)
+      }
       continue
     }
     fs.rmSync(path.resolve(dir, file), { recursive: true, force: true })
   }
+  if (nested.length) {
+    skip = skip
+      .filter((f) => path.dirname(f) !== '.')
+      .map((f) => f.replace(/.+?[\\/]/, ''))
+    for (const dir of nested) {
+      emptyDir(dir, skip)
+    }
+  }
```

Comment:
We can't use `startsWith` here, because we could have file `a` and skip `ab`. If we go with this approach we need to use `path.relative`. I was thinking we could preprocess the skip list to find what we should skip in this level first. Something like (not tested):
```js
function emptyDir(dir, skip) {
  let nested, skipInDir
  if (skip) {
    skipInDir = []
	  for (const file of skip) {
	    const dirname = path.dirname(file);
	    if (dirname !== '.') {
	      file.slice(dirname);
        nested ??= new Map()
        const nestedSkip = nested.get(dirname);
	      if (!nestedSkip) {
	        nestedSkip = [];
	        nested.set(dirname, nestedSkip);
	      }
	      if (!skipInDir.includes(dirname)) {
	        skipInDir.push(dirname);
	      }
	    } else {
	      skipInDir.push(file);
	    }
	  }
  }
  for (const file of fs.readdirSync(dir)) {
    if (skipInDir?.includes(file)) {
      continue;
    }
    const resolvedFile = path.resolve(dir, file);
    if (nested?.has(file)) {
      emptyDir(resolvedFile, nested.get(file));
    } else {
      fs.rmSync(path.resolve(dir, file), { recursive: true, force: true });
    }
  }
}
```

Reply:
My missing, there has one place in your demo codes that I can not understand, is `file.slice(dirname)`.

Reply:
Yes, sorry, it should be `dirname.length`, maybe +1

Reply:
I also thought the following way to improve `startsWith`, is it feasible?

```js
if (isWindows)
  skip = skip.map(slash)

// ...
const matched = skip.find((f) => f === file || f.startsWith(`${file}/`))
```

Reply:
Yes, I think that could also work 👍🏼 
You can use slash directly. But I think your logic for the nested skip may also need to be revised. It should depend on each nested dir, no?

Reply:
Yes, you are right. I revise it now.